### ISSUE TITLE
Laminography code crashes

### ISSUE THREAD

**Issue by cabouman** [cabouman/mbircone#143]
It has been observed that the laminography code crashes with the following parameters

Here is the function call the results in a crash
```
vol = recon_lamino(y, angles, lami_angle, num_image_rows=800,
                  num_image_cols=800, num_image_slices=120,
                  stop_threshold=0.05, snr_db=30,
                  sharpness=0.0, positivity=False)

```
Here are the parameters:

- y is a (361, 300, 600) array.
- Ubuntu VM with 8 cores (Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz) on an OSX host
- 16GB of RAM on computer

**Comment by DamonLee5** [comment]
I successfully reproduce this nan issue using this data size in branch lamino_crash.
If you want to reproduce the nan issue, you can just run demo_laminography_test.py

**Comment by cabouman** [comment]
Excellent! Next we need to figure out where the first nan is produced.

**Comment by DamonLee5** [comment]
I checked down the bug and found out that the first infinity value happen in theta1_f.

I print out the value of theta1_f before it became an infinity value
icdInfo[k_M].theta1_f = 2.993959e+38 

The max value of float is 3.402823466 E + 38
This thing keeps accumulating in line 961 and finally became an infinity value.
icdInfo[k_M].theta1_f += parallelAux->partialTheta[threadID][k_M].t1;

**Comment by cabouman** [comment]
Wenrui, good detective work. We are getting close. We may need to Zoom in order to track down the problem. It appears that a pixel with no associated sonogram measurements is for some reason accumulating an infinite value for the first derivative. It's probably some strange corner case that Thilo never thought about. If necessary, we might need to contact Thilo to see if he can give us some advice.

Comments/questions:
- This shows up on line 935 in icd3d.c in the limino_crash branch. Is that correct? Line 961 has something different.
- Is the associated value of theta2_f == 0 , exactly? If so, then that confirms that the pixel update that is causing the problems has no projections.
- Is the value of icdInfo[k_M].theta1_f overflowing the maximum float value of 3.402823466 E + 38?

**Comment by DamonLee5** [comment]
- This shows up on line 935 in icd3d.c in the limino_crash branch. Is that correct? Line 961 has something different.

Right. It is line 935. I add some printf function in my local branch, which I don't want to push to GitHub.

- Is the associated value of theta2_f == 0 , exactly? If so, then that confirms that the pixel update that is causing the problems has no projections.
No the associate value of theta2_f is not 0.
icdInfo[k_M].theta1_f 2.993959e+38
icdInfo[k_M].theta2_f 5.407720e+04

- Is the value of icdInfo[k_M].theta1_f overflowing the maximum float value of 3.402823466 E + 38?
Yes. I record previous value of icdInfo[k_M].theta1_f and do the accumulation. When detected icdInfo[k_M].theta1_f became infinity print its previous value. 

            qq = icdInfo[k_M].theta1_f;
            icdInfo[k_M].theta1_f += parallelAux->partialTheta[threadID][k_M].t1;

            icdInfo[k_M].theta2_f += parallelAux->partialTheta[threadID][k_M].t2;
            if isnan(icdInfo[k_M].theta1_f){
                printf( "icdInfo[k_M].theta1_f %e\n", qq);
                printf( "icdInfo[k_M].theta2_f %e\n", icdInfo[k_M].theta2_f);

                exit(-1);
            }
            if isinf(icdInfo[k_M].theta1_f){
                printf( "icdInfo[k_M].theta1_f %e\n", qq);
                printf( "icdInfo[k_M].theta2_f %e\n", icdInfo[k_M].theta2_f);
                exit(-1);
            }

**Comment by dyang37** [comment]
It seems that Wenrui's test script is not an exact match of Brendt's test case, and it crashed due to a different reason.
In Wenrui's script, we have `theta_degrees=361`. Instead, we should set `num_views=361` to reproduce Brendt's case.
In fact, Wenrui's test case crashed because we have this unusual tilt angle of 361 degree. If we change the tilt angle back to 60 degrees, then the nan problem goes away.

**Comment by DamonLee5** [comment]
According to Diyu's input, I change the tilt angle to 60 deg and set the num_views to 361. (demo_laminography_test2.py )

python demo_laminography_test2.py 
This script is a demonstration of the laminography reconstruction algorithm. Demo functionality includes 
	 * Generating a 3D laminography sample phantom; 
	 * Generating synthetic laminography data by forward projecting the phantom; 
	 * Performing a 3D qGGMRF reconstruction; 
	 * Displaying the results.
Generating a laminography phantom ...
Phantom shape is: 120 64 64
Generating synthetic sinogram ...
[ticToc] computeSysMatrix = 8.031289e+00 s

SystemMatrix parameters:
	i_vstride_max = 3 
	i_wstride_max = 2 
	N_u = 1082 
	Delta_u = 2.500000e-01 
	u_0 = -1.355571e+02 
	u_1 = 1.349429e+02 
	B_ij_max = 1.000001e+00 
	C_ij_max = 1.000001e+00 
	B_ij_scaler = 3.921571e-03 
	C_ij_scaler = 3.921572e-03 

Writing System Matrix to /Users/damonli/.cache/mbircone/sysmatrix/efb0f7ad8debef9ab8d9_pid6933_rndnum588.sysmatrix 
Total size written = 1.071125e-01 GB
Synthetic sinogram shape: (num_views, num_det_rows, num_det_channels) =  (361, 300, 600)
Performing 3D qGGMRF reconstruction ...
max_resolution =  2
Performing multi-resolution recon for reconstruction size (slices, rows, cols)=(60, 400,400).
Performing multi-resolution recon for reconstruction size (slices, rows, cols)=(30, 200,200).
[ticToc] computeSysMatrix = 8.481022e+00 s

SystemMatrix parameters:
	i_vstride_max = 6 
	i_wstride_max = 5 
	N_u = 1127 
	Delta_u = 1.000000e+00 
	u_0 = -5.648557e+02 
	u_1 = 5.621443e+02 
	B_ij_max = 5.644600e+00 
	C_ij_max = 1.154701e+00 
	B_ij_scaler = 2.213569e-02 
	C_ij_scaler = 4.528241e-03 

Writing System Matrix to /Users/damonli/.cache/mbircone/sysmatrix/04f44174595006822347_pid6933_rndnum889.sysmatrix 
Total size written = 1.591105e-01 GB

Image parameters read:
	x_0 = -4.000000e+02 
	y_0 = -4.000000e+02 
	z_0 = -1.385641e+08 
	N_x = 200 
	N_y = 200 
	N_z = 30 
	Delta_xy = 4.000000e+00 
	Delta_z = 4.000000e+00 
	j_xstart_roi = -1 
	j_ystart_roi = -1 
	j_zstart_roi = -1 
	j_xstop_roi = 0 
	j_ystop_roi = 0 
	j_zstop_roi = 0 

Sinogram parameters read:
	N_dv = 600,
	N_dw = 300,
	Delta_dv = 1.000000e+00,
	Delta_dw = 1.154701e+00,
	N_beta = 361,
	u_s = -2.400000e+08,
	u_r = 0.000000e+00,
	v_r = 0.000000e+00,
	u_d0 = 0.000000e+00,
	v_d0 = -3.000000e+02,
	w_d0 = -1.385642e+08,
	(potentially uninitialized:)
	weightScaler_value = -1.000000e+00,

Reconstruction parameters read:
	proximal map mode = 0 
	q = 2.000000e+00 
	p = 1.200000e+00 
	T = 1.000000e+00 
	sigmaX = 5.444920e-03 
	bFace = 1.666667e-01 
	bEdge = -1.000000e+00 
	bVertex = -1.000000e+00 
	sigma_lambda = 1.000000e+00 
	is_positivity_constraint = 0 
	stopThresholdChange_pct = 5.000000e-02 
	stopThesholdRWFE_pct = 0.000000e+00 
	stopThesholdRUFE_pct = 0.000000e+00 
	MaxIterations = 100 
	relativeChangeMode = meanImage 
	relativeChangeScaler = 1.000000e-01 
	relativeChangePercentile = 9.990000e+01 
	N_G = 2 
	zipLineMode = 2 
	numVoxelsPerZiplineMax = 200 
	numVoxelsPerZipline = 30 
	numZiplines = 1 
	weightScaler_estimateMode = None 
	weightScaler_domain = spatiallyInvariant 
	weightScaler_value = 6.903706e+00 
	NHICD_Mode = off 
	NHICD_ThresholdAllVoxels_ErrorPercent = 8.000000e+01 
	NHICD_percentage = 1.500000e+01 
	NHICD_random = 2.000000e+01 
	verbosity = 1 
	isComputeCost = 1 
************************** Iteration 0  (max. 100) **************************
*  Cost                   = 6.8732352000e+07
*  Rel. Update            = 0.0000000000e+00 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 1.0000000000e+02 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 1.0000000000e+02 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 1.4604738235e+01 = 1/0.0684709325
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 0.0000000000e+00 
*  time icd update        = 8.8499998674e-04 s
*  ratioUpdated           = 0.0000000000e+00 %
*  totalEquits            = 0.0000000000e+00 
******************************************************************************

************************** Iteration 1  (max. 100) **************************
*  Cost                   = 4.7409372000e+07
*  Rel. Update            = 1.0000000000e+02 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 8.2583580017e+01 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 8.2583580017e+01 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 9.9605007172e+00 = 1/0.1003965586
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 6.9725465335e+03 
*  time icd update        = 6.7610879000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 1.0000000000e+00 
******************************************************************************

************************** Iteration 2  (max. 100) **************************
*  Cost                   = 4.8311232000e+07
*  Rel. Update            = 8.6062286377e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 8.2465354919e+01 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 8.2465354919e+01 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 9.9320030212e+00 = 1/0.1006846279
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 1.3542701005e+04 
*  time icd update        = 3.4809895000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 2.0000000000e+00 
******************************************************************************

************************** Iteration 3  (max. 100) **************************
*  Cost                   = 6.2264956000e+07
*  Rel. Update            = 7.1252021790e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 9.3070571899e+01 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 9.3070579529e+01 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 1.2650816917e+01 = 1/0.0790462792
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 2.5948504477e+04 
*  time icd update        = 1.8167521000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 3.0000000000e+00 
******************************************************************************

************************** Iteration 4  (max. 100) **************************
*  Cost                   = 1.3392889600e+08
*  Rel. Update            = 7.7944404602e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 1.3586082458e+02 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 1.3586082458e+02 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 2.6957662582e+01 = 1/0.0370952040
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 2.6037960012e+04 
*  time icd update        = 1.8105106000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 4.0000000000e+00 
******************************************************************************

************************** Iteration 5  (max. 100) **************************
*  Cost                   = 3.3764943872e+10
*  Rel. Update            = 9.8799171448e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 2.2093190918e+03 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 2.2093190918e+03 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 7.1287060547e+03 = 1/0.0001402779
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 6.5853431316e+03 
*  time icd update        = 7.1586247000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 5.0000000000e+00 
******************************************************************************

************************** Iteration 6  (max. 100) **************************
*  Cost                   = 7.8212830855e+15
*  Rel. Update            = 9.9982681274e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 1.0667088750e+06 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 1.0667090000e+06 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 1.6618263040e+09 = 1/0.0000000006
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 1.7183480880e+04 
*  time icd update        = 2.7434489000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 6.0000000000e+00 
******************************************************************************

************************** Iteration 7  (max. 100) **************************
*  Cost                   = 2.5006110830e+22
*  Rel. Update            = 9.9996620178e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 1.9074030080e+09 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 1.9074030080e+09 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 5.3134753037e+15 = 1/0.0000000000
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 2.6707658218e+04 
*  time icd update        = 1.7651118000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 7.0000000000e+00 
******************************************************************************

************************** Iteration 8  (max. 100) **************************
*  Cost                   = 1.5879404979e+29
*  Rel. Update            = 9.9998764038e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 4.8065832550e+12 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 4.8065832550e+12 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 3.3741686732e+22 = 1/0.0000000000
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 1.7846225365e+04 
*  time icd update        = 2.6415671000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 8.0000000000e+00 
******************************************************************************

************************** Iteration 9  (max. 100) **************************
*  Cost                   = 1.1962385336e+36
*  Rel. Update            = 9.9997154236e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = 1.3192536437e+16 % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = 1.3192536437e+16 % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = 2.5418526717e+29 = 1/0.0000000000
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 1.9480488891e+04 
*  time icd update        = 2.4199598000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 9.0000000000e+00 
******************************************************************************

************************** Iteration 10 (max. 100) **************************
*  Cost                   = inf       
*  Rel. Update            = 9.9998825073e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = inf        % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = inf        % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = inf        = 1/0.0000000000
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 2.4912911560e+04 
*  time icd update        = 1.8922718000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 1.0000000000e+01 
******************************************************************************

************************** Iteration 11 (max. 100) **************************
*  Cost                   = inf       
*  Rel. Update            = 9.9998077393e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = inf        % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = inf        % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = inf        = 1/0.0000000000
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 1.9983404389e+04 
*  time icd update        = 2.3590576000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 1.1000000000e+01 
******************************************************************************

************************** Iteration 12 (max. 100) **************************
*  Cost                   = inf       
*  Rel. Update            = 9.9999053955e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = inf        % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = inf        % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = inf        = 1/0.0000000000
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 9.6365817743e+03 
*  time icd update        = 4.8919836000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 1.2000000000e+01 
******************************************************************************

************************** Iteration 13 (max. 100) **************************
*  Cost                   = inf       
*  Rel. Update            = 9.9999404907e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = inf        % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = inf        % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = inf        = 1/0.0000000000
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 1.6366534835e+04 
*  time icd update        = 2.8803898000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 1.3000000000e+01 
******************************************************************************

************************** Iteration 14 (max. 100) **************************
*  Cost                   = inf       
*  Rel. Update            = 9.9999458313e+01 % (threshold = 5.0000000745e-02 %)
*  RWFE = ||e||_W/||y||_W = inf        % (threshold = 0.0000000000e+00 %)
*  RUFE = ||e|| / ||y||   = inf        % (threshold = 0.0000000000e+00 %)
* ----------------------------------------------------------------------------
*  1/M ||e||^2_W          = inf        = 1/0.0000000000
*  weightScaler_value     = 6.9037060738e+00 = 1/0.1448497325
* ----------------------------------------------------------------------------
*  voxelsPerSecond        = 2.6376323817e+04 
*  time icd update        = 1.7872847000e+01 s
*  ratioUpdated           = 1.0000000000e+02 %
*  totalEquits            = 1.4000000000e+01 
******************************************************************************

icdInfo[k_M].theta1_f -3.228554e+38
icdInfo[k_M].theta2_f 5.410564e+04


The cost keeps increasing since Iteration 1

**Comment by dyang37** [comment]
A temporary fix was implemented in PR #152.
